### PR TITLE
added parser for "Studentenwerk Düsseldorf"

### PR DESCRIPTION
**Pull request by Athemis**
Parser for canteens operated by "Studentenwerk Düsseldorf". Food additive list is maintained as PDF and difficult to parse, hence a static dictionary compiled from that PDF is included with the parser.

### CODE REVIEW

File: duesseldorf.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+            # parse menu items
+            meal, raw_legend = parse_menu_items(items)
+            # Studentenwerk Düsseldorf does not maintain a easily readable list of additives etc.
+            # need to do our own parsing against a static list
+            legend = parse_legend(raw_legend)
+            # get prices for each role
+            prices = parse_prices(counter.find('ul', 'price'))
+            # add meal to list
+            canteen.addMeal(date, counter_name, meal, notes=legend, prices=prices)
+        i += 1
```
Comment (klemens) [inline duesseldorf.py]: This manual counting seems a little brittle. Why don't you iterate over the `counter` divs directly (and merge with the previous function)?

File: duesseldorf.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
+import re
+from urllib.request import urlopen
+
+from bs4 import BeautifulSoup as parse
+
+from pyopenmensa.feed import OpenMensaCanteen
+from utils import Parser
+
+# this will match notes, e.g. "[R]" or "(1,10,12)"
+legend_regex = re.compile('(?:[\[(]{1}[ARFSVZ\d\,]+[\])])')
```
Comment (klemens) [inline duesseldorf.py]: What is the purpose of the non-capturing group? `{1}` also seems redundant as it's the default anyway.

File: duesseldorf.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+                text = text.replace(li, '').strip()
+                # in addition, remove brackets left over after note deletion
+                for ch in ['(', ')', '[', ']']:
+                    li = li.replace(ch, '')
+                # split notes in list and append master list
+                li = li.split(',')
+                raw_legend.append(li)
+            strings.append(text)
+    # join all strings and return result + raw legend
+    return (", ".join(strings), raw_legend)
```
Comment (klemens) [inline duesseldorf.py]: Should we remove the ` - Preis ab:` and ` - Preis je 100g:` suffixes from salads and desserts? They seem redundant, as the meals also additionally contain a "Bitte beachten Sie die Auszeichnung am Produkt" note.

File: duesseldorf.py (hunk 1 above)
Comment (Athemis) [inline duesseldorf.py]: The problem here is that the counter's names (e.g. "Essen I", "Grill", etc.) are not part of that div, but are provided as separate h2 outside the counter divs. Hence I get two individual lists with no obvious way (at least not obvious to me, hints appreciated ;-) ) to link them to each other directly.

For reference, you may have a look at the [html source of the site itself](http://www.stw-d.de/gastronomie/speiseplaene/mensa-universitaetsstrasse-duesseldorf/).

File: duesseldorf.py (hunk 2 above)
Comment (Athemis) [inline duesseldorf.py]: Totally agreed and changed.

File: duesseldorf.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+            legend_tmp = legend_regex.findall(text)
+            for li in legend_tmp:
+                # remove notes from text to obtain a clean string
+                text = text.replace(li, '').strip()
+                # in addition, remove brackets left over after note deletion
+                for ch in ['(', ')', '[', ']']:
+                    li = li.replace(ch, '')
+                # split notes in list and append master list
+                li = li.split(',')
+                raw_legend.append(li)
```
Comment (Athemis) [inline duesseldorf.py]: Right. Completely overlooked the possibility of duplicate entries here. Thanks!

File: duesseldorf.py (hunk 3 above)
Comment (Athemis) [inline duesseldorf.py]: Agreed for "- Preis ab:". As for "- Preis je 100g:" I'm not sure if it's not better to keep this one as the shown prices are really only per 100 g of salad (you'll have to weight it before checkout)

File: duesseldorf.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+            canteen.addMeal(date, counter_name, meal, notes=legend, prices=prices)
+        i += 1
+
+
+def parse_legend(raw_legend):
+    verbose_legend = []  # we'll add parsed legend entries here
+    for li in raw_legend:  # raw_legends contains a two dimensional array
+        for li2 in li:  # second dimension
+            for key in legend_dict:  # iterate over keys and check against legend item (li2)
+                if li2 == key:
```
Comment (Athemis) [inline duesseldorf.py]: Changed

File: duesseldorf.py (hunk 1 above)
Comment (klemens) [inline duesseldorf.py]: I meant the div with the `counter` class instead fo the `counter-table` class:
```html
<div class="counter">
    <h2>Essen I</h2>
    <div class="counter-table">...</div>
</div>
```

Comment (Athemis) [inline duesseldorf.py]: Oops :-P ok, that was probably too obvious. Changed.

File: duesseldorf.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+               }
+
+# strings to be removed from meal title
+remove_strings = ['- Preis je 100g:',
+                  '- Preis ab:']
+
+
+def parse_url(url, today=False):
+    canteen = OpenMensaCanteen()
+    document = parse(urlopen(url).read())
```
Comment (klemens) [inline duesseldorf.py]: Looking at the commits by @mswart from today (0e39dc58a4a1d4cc9267f277eea225f1e3ef50d3), we should probably explicitly specify the parser here.

Comment (Athemis) [inline duesseldorf.py]: Done

**Comment by klemens** [comment]
I rebased, squashed and merged your changes. Thanks again!

I will now try to add the parser to the openmensa.org page.

**Comment by klemens** [comment]
All canteens are online now. I took the addresses from this page: http://www.stw-d.de/gastronomie/oeffnungszeiten/

**Pull request merged**